### ISSUE TITLE
Keyboard layout changes to wrong layout during usage

### ISSUE THREAD

**Issue by Temez1** [arnog/mathlive#1194]
## Description

https://user-images.githubusercontent.com/46352445/134766193-97266791-32df-4206-a94d-889fd194fdf2.mp4

Finnish keyboard layout: http://kbdlayout.info/KBDFI/

So initially the layout is all good. As you can see, I can type multiple "-" (next to right-shift) on my Finnish keyboard. But when I "submit" the answer the next time I'm using mathfield the layout has changed to a wrong layout, where pressing the same key leads it to be interpreted as "/" which leads to a fraction.

About that "submit". Copy-pasting all my code to codesandbox would take a bit too much time I try to explain the part of code that I think is relevant to this issue. I could share @arnog my private repo for a moment to debug this if you think that's needed.

```typescript
  const handleOnEnterKeyPressedOrFocusLostAndValueChanged = () => {
    if (mfe.value !== "") {
      // This is a long function (that comes from parent component) that get's called when I "submit" the answer
     // Eg. Checking the answer, clearing states, showing the success message etc.
      onEnterKeyPressedOrFocusLostAndValueChanged(mfe.getValue(format))
      mfe.setValue("")
    }
  }

  useLayoutEffect(() => {
    console.log("Mounting mathfield")

    ref.current?.appendChild(mfe)
    mfe.addEventListener(
      "change",
      handleOnEnterKeyPressedOrFocusLostAndValueChanged
    )
    mfe.focus()
    return () => {
      console.log("Unmounting mathfield")

      mfe.removeEventListener(
        "change",
        handleOnEnterKeyPressedOrFocusLostAndValueChanged
      )
      ref.current?.removeChild(mfe)
    }
  }, [])
```

## Steps to Reproduce

1. I start using mathfield and it's working
2. As long as I don't press "-" everything works as intended
3. When I at some point of time press "-", the next math challenge I try to type "-" is interpreted as a fraction

### Actual Behavior

I can't type "-"

### Expected Behavior

I can type "-"

### Environment

**MathLive version**

0.69.7

**Comment by arnog** [comment]
Is this the code sandbox that reproduces the issue? https://codesandbox.io/s/layout-bug-v9udw?file=/src/Mathfield.tsx
Have you tried reproducing the issue without Vue (i.e. using a plain web component)?

**Comment by Temez1** [comment]
Yes, that reproduces the issue when the mathfield is remounted. This happens also without remounting as described here, but due that the function call is so long, I haven't reproduced the issue without remounting in the sandbox.

Anyways, the behavior of the bug is exactly the same even if it's remounted or not. You can't type "-" on the keyboard that has the minus key next to the right shift (it works once before "submit" as described above). As the layout changes to the wrong layout for some reason.

That sandbox & I'm using React (not Vue). I haven't tried reproducing the issue using the plain web components.

I'm sorry if this is a bit unclear description. But I think the most important part of this bug is that it works first, but after refreshing (remounted or not) the layout changes to I believe English layout where the "/" is where in Finnish layout the "-" is (first button on the left of right shift). See the picture of the English layout below. Source [Wikipedia](https://en.wikipedia.org/wiki/British_and_American_keyboards)

![image](https://user-images.githubusercontent.com/46352445/134812562-d9f7df6a-1bc9-44b5-9e75-98b7d5175bb1.png)

**Comment by arnog** [comment]
OK, so you think that what triggers the switch to the incorrect keyboard layout is calling `setValue()`? Is that what you mean by "refreshing"?
If you had an example that demonstrated that issue with a plain web component, that would speed up my ability to debug it, but I'll try to see what I can do.

**Comment by Temez1** [comment]
No, it's not about setValue. I removed that and it still happens. Removed that also from the sandbox. There it clears anyways because it's remounted and the values don't stay in the mathfield (if you give it a try).

**Comment by arnog** [comment]
ok, so this *is* an issue about remounting? Or is it losing the focus? Or something else?

Regarding web components, they are very easy to use. It's literally one line of HTML. You can see an example here: https://cortexjs.io/mathlive/demo/ and perhaps you can modify this example to reproduce the problem you're experiencing? (not the source code of the example is editable)

**Comment by Temez1** [comment]
@arnog Please see my DM on Gitter. I don't want to clutter this comment section with different observations.

**Comment by Temez1** [comment]
@arnog Well, I add the list of things that I have observed here as well.

- Happens with web components as well (added to codesandbox)
- Happens when remounting (CodeSandbox version)
- Happens without remounting (local version)
- The local version is not even losing focus between the answers (referring to the video)
- The initial correct layout doesn't support fully a Finnish layout (eg. "^") but this is another issue
- The changed/wrong layout support ÖÄÅ characters and shortcuts, but the minus "-" doesn't work anymore.

So all in all, I have a hard time understanding what changes and why. But somehow when I press "-" it works for the first time and then triggers a change in the layout for the next answer.

Just let me know how you want to proceed to debug this.

**Comment by arnog** [comment]
Yeah, the Finnish keyboard not working completing is not surprising. There is no Finnish keyboard layout supported. I think the closest one is the German keyboard (you can check which keyboard layout is in action with `mfe.keyboardLayoutName` I think). Please feel free to submit a separate issue/PR about this.

I'll have a look at the updated code sandbox. This may be in a bit, I'm currently on the road with only a phone and poor connectivity.

And I hope you'll think of me when your startup becomes wildly successful 😀

**Comment by Temez1** [comment]
I still think after the discussion that instead of trying to support all different layouts (and also detect the right layout) some other approach would be more scalable & easier to handle than the layout approach. That would be quite a rework so I understand if that's not possible right now. I will look into that Finnish keyboard layout support later at some point when it's necessary (probably pretty soon).

No rush, I can keep doing other things meanwhile.

Haha, If that day ever comes :) I mean if I manage to get funding & resources for sure I would support this & OSS for making a great job. Unfortunately, currently, I'm solo so most of my effort goes to build the MVP so I can someday get those resources.

Nevertheless, thank you already for your kind help & support with this great library.

And if I can help with my small understanding let me know.

**Comment by arnog** [comment]
If you want to try to implement a different way of handling keyboard shortcuts I'd be more than happy to review a PR. Given that VSCode ran into the same issue and came up with the same solution, that's also something they would be interested in, I'm sure.

**Comment by Temez1** [comment]
I might take a look into that later in time. But now that's not on the priority list.

Anyways, I think right now the best solution is to have the possibility to set the layout manually, so I can provide the Finnish layout in some way & there is no way it changes to the wrong layout over time. Or what do you think is the solution?

**Comment by arnog** [comment]
The solution is two-fold: fix the bug that causes the layout to switch and add support for the Finnish keyboard layout.

I realized I may not have explained clearly enough why MathLive uses an apparently complicated approach to recognizing keyboard shortcuts.

Let's consider the keyboard shortcut Ctrl+Alt/Option-A.

With a US keyboard layout, the event received will have `code: 'KeyA', key: '\u00e5'`.
On a French AZERTY keyboard layout, the event received will have `code: 'KeyQ', key: '\u00e6`.

Why is the code `KeyQ`? On this keyboard layout, the Q and A keys are swapped. This is not unusual. While some keys retain their positions, many keys are swapped around or altogether unique in some layouts, particularly for punctuations and symbols. However, the `code` field of the event does not represent the label of the key, but indicates the physical position of the key as if it was on a US keyboard, in this case "the key immediately to the right of the caps lock key, which is labeled Q on a US keyboard (but is labeled A on a French keyboard".

What we want to know is that the user pressed a key labeled "A". But none of the information in the event record tells us that. The value of the `key` field varies depending on the keyboard layout and the modifiers pressed.

However, if we know the keyboard layout, we can use a table that maps the value of the `key` field to infer the label of the key (i.e. what the user sees printed on the top of the key cap) the user pressed, regardless of its physical location on the keyboard. Once we have the label, we can figure out that the user pressed Ctrl+Alt+A using the modifier fields of the event.

But how do we know what is the current keyboard layout? There is unfortunately no web platform API to obtain that information. So one approach is to indicate programmatically to MathLive which keyboard layout the user is using. Otherwise, MathLive will use the user locale to guess the keyboard (for example, guessing to use the French AZERTY keyboard if the user locale is France). 

Finally, MathLive uses a heuristic to refine its guess: with each keyboard event, MathLive checks that the info in the event record (specifically the `code` and `key` fields) are consistent with the current keyboard layout. If not, it finds a better matching keyboard layout, and will switch to that keyboard layout if it is confident enough of that guess.

MathLive currently has a limited set of "known" keyboard layouts. If you happen to use a keyboard layout MathLive doesn't know, it will guess the wrong keyboard layout. As a result some keyboard shortcuts may produce unexpected results.

Adding new keyboard layouts is not hard, but building the table mapping physical keys to characters is tedious and has been done so far based on user requests. Contributions of additional keyboard layouts are welcome.

That said, there seems to be one (or more?) issues where MathLive "forgets" the current keyboard layout, or possibly decides to change to another keyboard layout. Test cases that demonstrate this problem would be extremely useful to debug and fix this issue.

**Comment by Temez1** [comment]
Thanks for the explanation. However, it's a bit hard to follow at some point because it's partially either typoed or wrong. For example, the code is `KeyA` for both US & AZERTY layouts in your example. I'm using [this](https://developers.google.com/web/updates/2016/04/keyboardevent-keys-codes) as a reference.

> To give a concrete example, let’s assume your user is using a U.S. keyboard with a QWERTY keyboard layout. Pressing the physical Q key on that keyboard will result in a KeyboardEvent with a code attribute set to "KeyQ". This is true regardless of keyboard layout, and regardless of any other modifier keys. For comparison, on a French (AZERTY) keyboard this key would still have a code of "KeyQ" even though the letter printed on the keycap is an "a".  -[here](https://developers.google.com/web/updates/2016/04/keyboardevent-keys-codes#what%E2%80%99s_this_mean_in_practice)

So back to the idea that things could be simpler, I don't still understand what's the missing piece. If we want to create shortcuts that are the same regardless of the keyboard layout we would use the `code` field. I don't know when this would be the case, because I personally think that shortcuts are useful only because they have some intuition behind them. Eg. `ctrl + a` means select all, thus `key` being `a`.  I can't think of a use case where using `code` would be useful in a math context. But if you have some special combo that needs the `code` field let me know so I could handle that case in the pseudocode.

So with this info in mind, we would end up using the `key` for shortcuts & actual characters. And after this, the pseudocode would still look like this to detect what the user wanted to do with the keyboard regardless of what layout is being used.

```
const key = KeyboardEvent.key 

// This would check if shift, ctrl, or alt was used and returns for example object with those fields being true or false
// I don't see that we need to know the order of modifiers being pressed.
const modifierKeyObject = getModifierKeyObject()  // returns object or null if no modifier keys were pressed
if (modifierKeyObject is not null){
    doShortcut(modifierKey, key)
    return
  }
}
doSomethingWithKey(key)
```

I played with [this](https://w3c.github.io/uievents/tools/key-event-viewer.html) tool & things seem to be correct with this logic.

**Comment by arnog** [comment]
I suggest you re-read the paragraph you quoted. It is correct: on a French AZERTY keyboard, the key labeled "A" has a code of KeyQ. Hence the problem.

**Comment by Temez1** [comment]
Re-read it and it makes sense now. However, I don't get this part

> What we want to know is that the user pressed a key labeled "A". But none of the information in the event record tells us that. The value of the key field varies depending on the keyboard layout and the modifiers pressed.

I don't see how the `KeyboardEvent.key` used and described above doesn't do exactly what you say here. The value is the key user pressed as it's labeled on the keyboard.

And as I said

>  I can't think of a use case where using `code` would be useful in a math context. But if you have some special combo that needs the code field let me know so I could handle that case in the pseudocode.

So, what exactly is the problem here?

**Comment by Temez1** [comment]
I installed the FRA language package on my windows and did that ctrl-alt-A with the US and French keyboards. I don't see any problem?

https://user-images.githubusercontent.com/46352445/135041682-c4efe7a5-ab35-4f6b-a198-fac68034b2ba.mp4

**Comment by stefnotch** [comment]
I took a quick look at this and it looks like there are new APIs being added, which might solve some of the issues. Namely [`Keyboard.getLayoutMap()`](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/getLayoutMap) and  [`Keyboard.lock()`](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/lock)

Also, the "keyboard layout" detecting code should probably not affect normal input, such as typing a `-` or `/`. It seems like it's only needed for shortcuts. At least that's what I would hope?

**Comment by arnog** [comment]
@stefnotch yes, maybe someday there will be standard API for this... At least "/" is considered a shortcut for "insert a fraction". It's actually a complex command as it has to consider if there is a selection, or a potential numerator to the left of the insertion point. I'm not sure about "-", I would have to check the code.

**Comment by Temez1** [comment]
@arnog Could you explain what's the problem in my last answer? As you see the key is `a` on both layouts. You said it's different but it isn't as you can see?

**Comment by arnog** [comment]
@Temez1 like I've said before, the value of the `key` property depends on the modifiers, keyboard layout and operating system. It's convenient that on your setup `alt+a` (or perhaps `ctrl+alt+a`) produces a value of `a`, but this is not something that can be relied on. As an example, here is what it produces on a Mac, with a US and French keyboard layout.

<img width="1281" alt="Screen Shot 2021-09-28 at 9 03 50 AM" src="https://user-images.githubusercontent.com/760798/135124355-a6c4b651-ebaf-44a8-b1f3-14fc5a721d96.png">
<img width="1281" alt="Screen Shot 2021-09-28 at 9 03 13 AM" src="https://user-images.githubusercontent.com/760798/135124370-a48248a3-1e7f-4005-b3c6-23a8334354ca.png">


At this point I've felt I've spent more time defending the implementation of this feature than the time it took me to implement it!

I will also reiterate that other open source projects that have attempted to solve this have implemented very similar solutions, including Visual Studio Code. Go review their implementation, it's instructive.

I suggest the following next steps. Either:

- submit a PR implementing a simpler solution. Make sure to test it on both Mac, Windows, Linux, iOS and Android. Test it with physical keyboards including a UK keyboard (it has a special "backslash" key which is found on no other keyboard), a French (or German) keyboard and a Japanese or Korean keyboard. Make sure that all the shortcuts work in all those configurations. I'd be happy to review and consider the PR for merging.

or

- submit a PR to support the Finnish keyboard layout (and any other keyboard layout you'd like to see supported)

**Comment by Temez1** [comment]
> _**NOTE**_ to tkainrad if you happen to read this. I just wanted to tag you to let you know that as you said this problem seems to be growing day by day. Maybe it should be solved somehow, we are living in 2021 anyways. :D

I agree we all have spent too much on this keyboard input thing. And that only speaks for itself how broken things are :D And I'm sorry if I have gone too far. I like to get to bottom of things to understand what is happening and why. And I understand that other projects have tried to solve this, but it seems like even [Google](https://webapps.stackexchange.com/questions/116405/google-spreadsheet-equivalent-of-excels-ctrlsemicolon-on-international-keyboar) can't solve this properly. So it doesn't make other solutions good.

And I think it would be a separate discussion anyways to start thinking about better alternatives. But I think, for now, that's not on the priority list. I might take a look into that later and open some kind of RFC if I have time. I already found a very good [article](https://tkainrad.dev/posts/why-keyboard-shortcuts-dont-work-on-non-us-keyboard-layouts-and-how-to-fix-it/) made by @tkainrad with some ideas to make the shortcuts international. 

I would be happy to make that PR if you give me simple instructions on where I should put that layout and what is needed on my part.

And what comes to the original bug, I think that's still an issue & it persists even if I add the Finnish layout. Let me know if I can help somehow with this bug.

**Comment by tkainrad** [comment]
Thansk for the mentin @Temez1. This was a very interesting read for me, as I have come to similar conclusions as @arnog over the past two years. Would have been very helpful to read these comments before that time.

Unfortunately, I don't have much to add to the discussion. @stefnotch already mentions [Keyboard.getLayoutMap()](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/getLayoutMap), which can be used to get a mapping from a key's universal `code` to the layout-dependent character that's printed on the user's keyboard. [KeyCombiner](https://keycombiner.com/collections/vscode/) uses this to display a layout-specific virtual keyboard for each user. However, the API is only available on Chromium-based browsers and has other problems, too.

In my experience, letting users define their own bindings is the only approach that kinda solves this problem for every possible layout and special case. They can just assign what works for them on their machine. Other approaches that I mentioned in my article (using `keypress`, the event's `key` property,...) could work well for simple apps, but make it very complicated when trying to provide support for many special characters, like I think MathLive is doing.

**Comment by Temez1** [comment]
> And what comes to the original bug, I think that's still an issue. Let me know if I can help somehow with this bug.

@arnog  Any update to this? This is a major bug. I can't use eg. negative numbers as long as this happens.

What comes to the Finnish layout, I take my words back. The "^" is only not working when it changes to the wrong layout. Initially, it does work as expected with the right layout. So right now there is no need to add a new layout, haven't found a use case where it's not working, yet.

**Comment by arnog** [comment]
@Temez1 nothing to report

**Comment by Temez1** [comment]
What's the plan to solve this issue/how to go forward?
